engine/src/asset/texture/mod.rs

```diff
@@ -0,0 +1,152 @@
+                .map(|(t, o)| {
+                    if o.animation_frames == 0 {
+                        TextureAnimationKind::Static(TextureMipKind::Mipped(t.clone()))
+                    } else {
+                        let mut frames = Vec::with_capacity(o.animation_frames as usize);
+                        let mut next = (t, *o);
+
+                        for i in 0..o.animation_frames {
+                            frames.push(TextureMipKind::Mipped(next.0.clone()));
+                            let (t, o) = &textures[next.1.next_animation_texture_id as usize];
+                            next = (t, o);
+                        }
+
+                        TextureAnimationKind::Animated(frames)
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            Ok((&[], textures))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        asset::color_map::{ColorMap, PaletteTexture},
+        utils::{format::*, test::*},
+    };
+    use std::{cell::LazyCell, path::PathBuf};
+
+    const COLOR_MAP_DATA: LazyCell<Vec<u8>> = deflated_file!("4F.dat");
+    const TEXTURE_INFO_DATA: LazyCell<Vec<u8>> = deflated_file!("93.dat");
+    const TEXTURE_DATA: LazyCell<Vec<u8>> = deflated_file!("95.dat");
+
+    #[test]
+    #[ignore = "uses Ashen ROM files"]
+    fn parse_rom_asset() -> eyre::Result<()> {
+        let (_, color_map) = <ColorMap as AssetParser<Pack>>::parser(())(&COLOR_MAP_DATA)?;
+        let color_map = &color_map.shades[15];
+        let (_, offsets) =
+            <TextureOffsetCollection as AssetParser<Pack>>::parser(())(&TEXTURE_INFO_DATA)?;
+        let (_, textures) =
+            <MippedTextureCollection as AssetParser<Pack>>::parser(&offsets)(&TEXTURE_DATA)?;
+
+        let output_dir = PathBuf::from(parsed_file_path!("textures/"));
+
+        textures
+            .iter()
+            .enumerate()
+            .try_for_each(|(i, texture)| match texture {
+                TextureAnimationKind::Static(TextureMipKind::NonMipped(_)) => {
+                    unreachable!("World textures are always mipped")
+                }
+                TextureAnimationKind::Static(TextureMipKind::Mipped(t)) => {
+                    t.mips.iter().enumerate().try_for_each(|(m, mip)| {
+                        let file = &output_dir.join(format!("{i:0>3X}-mip-{m}.png"));
+                        output_file(file, mip.colors.with_palette(color_map).to_png())
+                    })
+                }
+                TextureAnimationKind::Animated(t) => {
+                    let frames: Vec<_> = t
+                        .iter()
+                        .map(|t| match t {
+                            TextureMipKind::NonMipped(_) => {
+                                unreachable!("World textures are always mipped")
+                            }
+                            TextureMipKind::Mipped(t) => t,
+                        })
+                        .collect();
```

Comment:
nit: needless `collect`. (really interesting that clippy didn't picked this) [^1].

[^1]: I think it is because we are inside tests, so this doesn't _really_ matter.